plugins/modem-manager/fu-mm-device.c

```diff
@@ -684,6 +744,202 @@ fu_mm_device_write_firmware_qmi_pdc (FuDevice *device, GBytes *fw, GArray **acti
+	g_autoptr(XbBuilderSource) source = xb_builder_source_new ();
+	g_autoptr(XbSilo) silo = NULL;
+	g_autoptr(GPtrArray) components = NULL;
+	g_autoptr(GPtrArray) guids = NULL;
+
+	/* decompress entire archive ahead of time */
+	archive = fu_archive_new (fw, FU_ARCHIVE_FLAG_IGNORE_PATH, error);
+	if (archive == NULL)
+		return FALSE;
+
+	locker = fu_device_locker_new_full (device,
+					    (FuDeviceLockerFunc) fu_mm_device_mbim_open,
+					    (FuDeviceLockerFunc) fu_mm_device_mbim_close,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* get version_release */
+	data = fu_archive_lookup_by_fn (archive, "versionfile.xml", error);
+	if (data == NULL)
+		return FALSE;
+	if (!xb_builder_source_load_bytes (source, data,
+					   XB_BUILDER_SOURCE_FLAG_NONE, error))
+		return FALSE;
+	xb_builder_import_source (builder, source);
+	silo = xb_builder_compile (builder, XB_BUILDER_COMPILE_FLAG_NONE, NULL, error);
+	if (silo == NULL)
+		return FALSE;
+
+	guids = fu_device_get_guids (device);
+	components = xb_silo_query (silo, "components/component", 0, error);
+	if (components == NULL)
+		return FALSE;
+	for (guint i = 0; i < components->len; i++) {
+		XbNode *component = g_ptr_array_index (components, i);
+		const gchar *guid = xb_node_get_attr (component, "guid");
+		for (guint j = 0; j < guids->len; j++) {
+			if (g_strcmp0 (guid, g_ptr_array_index (guids, j)) == 0) {
+				version_release = xb_node_get_attr (component, "version");
+				break;
+			}
+		}
+		if (version_release)
+			break;
+	}
+
+	/* load the manifest of operations */
+	data = fu_archive_lookup_by_fn (archive, "flashfile.xml", error);
+	if (data == NULL)
+		return FALSE;
+	if (!xb_builder_source_load_bytes (source, data,
+					   XB_BUILDER_SOURCE_FLAG_NONE, error))
+		return FALSE;
+	xb_builder_import_source (builder, source);
+	silo = xb_builder_compile (builder, XB_BUILDER_COMPILE_FLAG_NONE, NULL, error);
+	if (silo == NULL)
+		return FALSE;
+
+	part = xb_silo_query_first (silo, "parts/part", error);
+	if (part == NULL)
+		return FALSE;
+	filename = xb_node_get_attr (part, "filename");
+	csum = xb_node_get_attr (part, "MD5");
+	data = fu_archive_lookup_by_fn (archive, filename, error);
+	if (data == NULL)
+		return FALSE;
+	csum_actual = g_compute_checksum_for_bytes (G_CHECKSUM_MD5, data);
+	if (g_strcmp0 (csum, csum_actual) != 0) {
+		g_debug ("[%s] MD5 not matched", filename);
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "[%s] MD5 not matched", filename);
+		return FALSE;
+	} else {
+		g_debug ("[%s] MD5 matched", filename);
+	}
+
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	fu_mbim_qdu_updater_write (self->mbim_qdu_updater, filename, data, device, error);
+	if (!fu_device_locker_close (locker, error))
+		return FALSE;
+
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	version = fu_mm_device_get_firmware_version_mbim (device, error);
+	if (g_strcmp0 (version, version_release) != 0) {
```

Comment:
Please see the comment in `fu_mbim_qdu_updater_caps_query_ready()` about this version string comparison. I don't think you need this check here, at this point it's enough to ensure the communication via MBIM is possible, and then say the upgrade was successful. fwupd will itself check and compare the version string in the CAB file w.r.t. the one reported by ModemManager after the device is uninhibited.